src/fips.c

```diff
@@ -271,44 +271,46 @@ fips_init(void)
 		fips_flag = '1';
 
 	if (fips_flag - '0') {
-#if !OPENSSL_VERSION_PREREQ(3, 0)
 		/* Set libica into FIPS mode. */
 		fips |= ICA_FIPS_MODE;
+	}
+}
 
+void
+fips_init(void)
+{
+	if (fips & ICA_FIPS_MODE) {
+#if !OPENSSL_VERSION_PREREQ(3, 0)
 		/* Try to set OpenSSL into FIPS mode. If this is not possible,
 		 * all software fallbacks (including RSA key generation) will
 		 * be disabled. OpenSSL FIPS mode can be queried using the
 		 * FIPS_mode() function. */
 		FIPS_mode_set(1);
 #else
-		fips = 0;
-
 #ifndef NO_FIPS_CONFIG_LOAD
 		/* Allow to skip reading the openssl 3.x fips config. Tests showed
 		 * that this step must be skipped on RHEL9 systems. But on other
 		 * systems, or with a locally built openssl, this step is necessary. */
 		if (!OSSL_LIB_CTX_load_config(openssl_libctx, LIBICA_FIPS_CONFIG)) {
 			syslog(LOG_ERR, "Libica failed to load openssl fips config %s\n",
 					LIBICA_FIPS_CONFIG);
-			fips |= ICA_FIPS_INTEGRITY;
+			fips = ICA_FIPS_INTEGRITY;
```

Comment:
Why did you change the |= into = ? The fips variable could contain the ICA_FIPS_MODE flag from the new fips_get_indicator function. So we would overwrite this here. However, in case of ICA_FIPS_INTEGRITY the library would stop working anyway.

Reply:
The old code did set fips = 0 in line 284.  Then in case of an error it or'ed ICA_FIPS_INTEGRITY onto it. The result is fips = ICA_FIPS_INTEGRITY in error case.
At the end of the if block (good case), it or'ed ICA_FIPS_MODE onto it.
So in good case fips = ICA_FIPS_MODE and in bad case fips = ICA_FIPS_INTEGRITY .

That whats the new code does, too, to, just less complicated :-) 

On entry of the block fips =ICA_FIPS_MODE, and in case of error I set it to fips = ICA_FIPS_INTEGRITY.
If no error occurred, it remains at fips =ICA_FIPS_MODE,